pkg/capacity/controller.go

```diff
@@ -0,0 +1,506 @@
+	if !found {
+		// The item was removed in the meantime. This can happen when the storage class
+		// or the topology segment are gone.
+		klog.V(5).Infof("Capacity Controller: %v became obsolete", item)
+		return nil
+	}
+
+	sc, err := c.scInformer.Lister().Get(item.storageClassName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			// Another indication that the value is no
+			// longer needed.
+			return nil
+		}
+		return fmt.Errorf("retrieve storage class for %+v: %v", item, err)
+	}
+
+	req := &csi.GetCapacityRequest{
+		Parameters: sc.Parameters,
+		// The assumption is that the capacity is independent of the
+		// capabilities. The standard makes it mandatory to pass something,
+		// therefore we pick something rather arbitrarily.
+		VolumeCapabilities: []*csi.VolumeCapability{
+			{
+				AccessType: &csi.VolumeCapability_Mount{},
+				AccessMode: &csi.VolumeCapability_AccessMode{
+					Mode: csi.VolumeCapability_AccessMode_UNKNOWN,
+				},
+			},
+		},
+	}
+	if item.segment != nil {
+		req.AccessibleTopology = &csi.Topology{
+			Segments: item.segment.GetLabelMap(),
+		}
+	}
+	resp, err := c.csiController.GetCapacity(ctx, req)
+	if err != nil {
+		return fmt.Errorf("CSI GetCapacity for %+v: %v", item, err)
+	}
+
+	quantity := resource.NewQuantity(resp.AvailableCapacity, resource.BinarySI)
+	if capacity != nil {
+		// Update existing object.
+		capacity.Capacity = quantity
+		var err error
+		klog.V(5).Infof("Capacity Controller: updating %s for %+v, new capacity %v", capacity.Name, item, quantity)
+		capacity, err = c.client.StorageV1alpha1().CSIStorageCapacities().Update(ctx, capacity, metav1.UpdateOptions{})
+		if err != nil && apierrs.IsConflict(err) {
+			// Handle the case where we had a stale copy of the object. Can only happen
+			// when someone else was making changes to it, which should be rare.
+			capacity, err = c.client.StorageV1alpha1().CSIStorageCapacities().Get(ctx, capacity.Name, metav1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("getting fresh copy of CSIStorageCapacity for %+v: %v", item, err)
+			}
+			capacity.Capacity = quantity
+			capacity, err = c.client.StorageV1alpha1().CSIStorageCapacities().Update(ctx, capacity, metav1.UpdateOptions{})
+		}
+		if err != nil {
+			return fmt.Errorf("update CSIStorageCapacity for %+v: %v", item, err)
+		}
+	} else {
+		// Create new object.
+		capacity = &storagev1alpha1.CSIStorageCapacity{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName:    "csisc-",
+				OwnerReferences: []metav1.OwnerReference{c.owner},
+			},
+			StorageClassName: item.storageClassName,
+			NodeTopology:     item.segment.GetLabelSelector(),
+			Capacity:         quantity,
+		}
+		if c.enumerateObjects >= 0 {
+			// Workaround for testing with a fake client: it doesn't
+			// set the name, so we have to make up something ourselves.
+			c.enumerateObjects++
+			capacity.Name = fmt.Sprintf("csisc-test-%d", c.enumerateObjects)
+			capacity.GenerateName = ""
+		}
```

Comment:
I don't like testing code in the controller. Use a reactor to generate name in controller_test.go, as suggested in https://github.com/kubernetes/client-go/issues/439 and implemented in https://github.com/bitnami-labs/sealed-secrets/pull/301

Reply:
I didn't like it either, but didn't know how to get name generation working. Thanks for the pointer, I'll add that.

Reply:
Done.